### ISSUE TITLE
Hyperv-vmcx fails trying to modify device 'Memory'

### ISSUE THREAD

**Issue by xalaliah** [hashicorp/packer#8633]
#### Overview of the Issue

I'm trying to create a VM with the Hyperv-vmcx builder, but keep getting the below error:
```
2020/01/20 16:35:34 ui error: ==> hyperv-vmcx: Error cloning virtual machine: PowerShell error: Hyper-V\Set-VMMemory : Failed to modify device 'Memory'.
==> hyperv-vmcx: 'Win10Ent2019LTSC_GOLD' failed to modify device 'Memory'. (Virtual machine ID 38C534CD-4FB7-485F-8632-D15D78C753E8)
==> hyperv-vmcx: At C:\Users\usx00014\AppData\Local\Temp\2\powershell147704452.ps1:44 char:1
==> hyperv-vmcx: + Hyper-V\Set-VMMemory -VM $compatibilityReport.VM -StartupBytes $memor ...
==> hyperv-vmcx: + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
==> hyperv-vmcx:     + CategoryInfo          : InvalidOperation: (:) [Set-VMMemory], VirtualizationException
==> hyperv-vmcx:     + FullyQualifiedErrorId : InvalidState,Microsoft.HyperV.PowerShell.Commands.SetVMMemory
```
I believe this is because the virtual machine is in a 'Saved' state at the time that this command runs, but I'm not entirely sure. When the script fails out, the VM is left in MMC in a saved state, hence my assumption. I have dynamic memory disabled and plenty of available RAM on the host. I also tried hardcoding the memory value to 4196 int instead of string, but same error. The vmcx path is pointing to the export directory, not directly at the vmcx or Virtual Machines or Virtual Hard Disks folders.

Edit: This also occurs if I omit the memory option completely and let it default to the 1028mb.

#### Reproduction Steps

Try to build image from vmcx path, modifying the memory from 1028MB to 4196MB.

### Packer version

`Packer v1.5.1` 

### Simplified Packer Buildfile

```
{
  "variables": {
    "vm_name": "Windows10Ent_01",
    "ram": "4196",
    "cpu": "2",
    "admin_password": "",
    "domain_username": "",
    "domain_password": "",
    "unattend_folder": "Win10Ent2019LTSC_GOLD",
    "vlan_id": "none",
    "scripts_dir": "\\\\devops-hyperv-2\\VDI\\Golden_Images\\Desktop\\Packer\\scripts",
    "os_ver": "",
    "collection_name": "{{user `group_name`}}_{{user `os_ver`}}",
    "group_name": "",
    "users_file": "\\\\devops-hyperv-2\\VDI\\Golden_Images\\Desktop\\Packer\\user_lists\\userlist.txt"
  },

  "builders": [
    {
      "vm_name": "{{user `vm_name`}}",
      "boot_wait": "6m",
      "communicator": "winrm",
      "cpus": "{{user `cpu`}}",
      "memory": "{{user `ram`}}",
      "winrm_username": "Admin",
      "winrm_password": "{{user `admin_password`}}",
      "winrm_timeout": "4h",
      "type": "hyperv-vmcx",
      "clone_from_vmcx_path": "\\\\devops-hyperv-2\\VDI\\Golden_Images\\Desktop\\Win10Ent2019LTSC_GOLD",
      "switch_name": "vSwitch",
      "vlan_id": "{{user `vlan_id`}}",
      "keep_registered": "true",
      "output_directory": "\\\\devops-hyperv-2\\VDI\\Instantiated_Images\\Desktop\\{{user `group_name`}}\\{{user `vm_name`}}",
      "shutdown_command": "{{user `scripts_dir`}}\\ShutdownForSysprep.bat {{user `vm_name`}} {{user `admin_password`}} {{user `domain_username`}} {{user `domain_password`}} {{user `unattend_folder`}}"
    }
  ],

```

### Operating system and Environment details

Host is Windows Server 2019 Datacenter. 
The VM I'm trying to spin up is Windows 10 Enterprise 2019 LTSC.

### Log Fragments and crash.log files

```
2020/01/21 10:09:05 [INFO] Packer version: 1.5.1 [go1.13.5 windows amd64]
2020/01/21 10:09:05 Attempting to open config file: C:\Users\usx00014\AppData\Roaming\packer.config
2020/01/21 10:09:05 [WARN] Config file doesn't exist: C:\Users\usx00014\AppData\Roaming\packer.config
2020/01/21 10:09:05 Setting cache directory: C:\Users\usx00014\packer_cache
cannot determine if process is in background: Process background check error: not implemented yet
2020/01/21 10:09:05 Plugin could not be found at C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-hyperv-vmcx (exec: "C:\\Users\\usx00014\\Downloads\\packer_1.5.1_windows_amd64\\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-hyperv-vmcx": file does not exist). Checking same directory as executable.
2020/01/21 10:09:05 Current exe path: C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe
2020/01/21 10:09:05 Creating plugin client for path: C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe
2020/01/21 10:09:05 Starting plugin: C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe []string{"C:\\Users\\usx00014\\Downloads\\packer_1.5.1_windows_amd64\\packer.exe", "plugin", "packer-builder-hyperv-vmcx"}
2020/01/21 10:09:05 Waiting for RPC address for: C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe
2020/01/21 10:09:06 packer.exe plugin: [INFO] Packer version: 1.5.1 [go1.13.5 windows amd64]
2020/01/21 10:09:06 packer.exe plugin: Attempting to open config file: C:\Users\usx00014\AppData\Roaming\packer.config
2020/01/21 10:09:06 packer.exe plugin: [WARN] Config file doesn't exist: C:\Users\usx00014\AppData\Roaming\packer.config
2020/01/21 10:09:06 packer.exe plugin: Setting cache directory: C:\Users\usx00014\packer_cache
2020/01/21 10:09:06 packer.exe plugin: args: []string{"packer-builder-hyperv-vmcx"}
2020/01/21 10:09:06 packer.exe plugin: Plugin port range: [10000,25000]
2020/01/21 10:09:06 packer.exe plugin: Plugin address: tcp 127.0.0.1:10000
2020/01/21 10:09:06 packer.exe plugin: Waiting for connection...
2020/01/21 10:09:06 Received tcp RPC address for C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe: addr is 127.0.0.1:10000
2020/01/21 10:09:06 packer.exe plugin: Serving a plugin connection...
2020/01/21 10:09:06 Plugin could not be found at C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-windows-shell (exec: "C:\\Users\\usx00014\\Downloads\\packer_1.5.1_windows_amd64\\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-windows-shell": file does not exist). Checking same directory as executable.
2020/01/21 10:09:06 Current exe path: C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe
2020/01/21 10:09:06 Creating plugin client for path: C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe
2020/01/21 10:09:06 Starting plugin: C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe []string{"C:\\Users\\usx00014\\Downloads\\packer_1.5.1_windows_amd64\\packer.exe", "plugin", "packer-provisioner-windows-shell"}
2020/01/21 10:09:06 Waiting for RPC address for: C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe
2020/01/21 10:09:06 packer.exe plugin: [INFO] Packer version: 1.5.1 [go1.13.5 windows amd64]
2020/01/21 10:09:06 packer.exe plugin: Attempting to open config file: C:\Users\usx00014\AppData\Roaming\packer.config
2020/01/21 10:09:06 packer.exe plugin: [WARN] Config file doesn't exist: C:\Users\usx00014\AppData\Roaming\packer.config
2020/01/21 10:09:06 packer.exe plugin: Setting cache directory: C:\Users\usx00014\packer_cache
2020/01/21 10:09:06 packer.exe plugin: args: []string{"packer-provisioner-windows-shell"}
2020/01/21 10:09:06 packer.exe plugin: Plugin port range: [10000,25000]
2020/01/21 10:09:06 packer.exe plugin: Plugin address: tcp 127.0.0.1:10000
2020/01/21 10:09:06 packer.exe plugin: Waiting for connection...
2020/01/21 10:09:06 Received tcp RPC address for C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe: addr is 127.0.0.1:10000
2020/01/21 10:09:06 packer.exe plugin: Serving a plugin connection...
2020/01/21 10:09:06 Plugin could not be found at C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-powershell (exec: "C:\\Users\\usx00014\\Downloads\\packer_1.5.1_windows_amd64\\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-powershell": file does not exist). Checking same directory as executable.
2020/01/21 10:09:06 Current exe path: C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe
2020/01/21 10:09:06 Creating plugin client for path: C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe
2020/01/21 10:09:06 Starting plugin: C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe []string{"C:\\Users\\usx00014\\Downloads\\packer_1.5.1_windows_amd64\\packer.exe", "plugin", "packer-provisioner-powershell"}
2020/01/21 10:09:06 Waiting for RPC address for: C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe
2020/01/21 10:09:06 packer.exe plugin: [INFO] Packer version: 1.5.1 [go1.13.5 windows amd64]
2020/01/21 10:09:06 packer.exe plugin: Attempting to open config file: C:\Users\usx00014\AppData\Roaming\packer.config
2020/01/21 10:09:06 packer.exe plugin: [WARN] Config file doesn't exist: C:\Users\usx00014\AppData\Roaming\packer.config
2020/01/21 10:09:06 packer.exe plugin: Setting cache directory: C:\Users\usx00014\packer_cache
2020/01/21 10:09:06 packer.exe plugin: args: []string{"packer-provisioner-powershell"}
2020/01/21 10:09:06 packer.exe plugin: Plugin port range: [10000,25000]
2020/01/21 10:09:06 packer.exe plugin: Plugin address: tcp 127.0.0.1:10000
2020/01/21 10:09:06 packer.exe plugin: Waiting for connection...
2020/01/21 10:09:06 Received tcp RPC address for C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe: addr is 127.0.0.1:10000
2020/01/21 10:09:06 packer.exe plugin: Serving a plugin connection...
2020/01/21 10:09:06 ui: Debug mode enabled. Builds will not be parallelized.
2020/01/21 10:09:06 ui: hyperv-vmcx: output will be in this color.
2020/01/21 10:09:06 ui: 
2020/01/21 10:09:06 Build debug mode: true
2020/01/21 10:09:06 Force build: false
2020/01/21 10:09:06 On error: 
2020/01/21 10:09:06 Preparing build: hyperv-vmcx
2020/01/21 10:09:06 packer.exe plugin: DiskBlockSize: 32
2020/01/21 10:09:06 packer.exe plugin: RamSize: 4196
2020/01/21 10:09:07 Debug enabled, so waiting for build to finish: hyperv-vmcx
2020/01/21 10:09:07 Starting build run: hyperv-vmcx
2020/01/21 10:09:07 Running builder: hyperv-vmcx
2020/01/21 10:09:07 [INFO] (telemetry) Starting builder hyperv-vmcx
2020/01/21 10:09:07 packer.exe plugin: Enter method: verifyPSVersion
2020/01/21 10:09:07 packer.exe plugin: $host.version.Major output: 5
2020/01/21 10:09:07 packer.exe plugin: Enter method: verifyPSHypervModule
2020/01/21 10:09:08 packer.exe plugin: Enter method: verifyHypervPermissions
2020/01/21 10:09:08 ui: ==> hyperv-vmcx: Creating build directory...
2020/01/21 10:09:08 packer.exe plugin: Created build directory: C:\Users\usx00014\AppData\Local\Temp\2\hyperv413148475
2020/01/21 10:09:08 ui: ask: ==> hyperv-vmcx: Pausing after run of step 'StepCreateBuildDir'. Press enter to continue.
2020/01/21 10:12:39 ui: ask: ==> hyperv-vmcx: Pausing after run of step 'StepOutputDir'. Press enter to continue.
2020/01/21 10:12:39 packer.exe plugin: No floppy files specified. Floppy disk will not be made.
2020/01/21 10:12:39 ui: ask: ==> hyperv-vmcx: Pausing after run of step 'StepCreateFloppy'. Press enter to continue.
2020/01/21 10:12:40 ui: ask: ==> hyperv-vmcx: Pausing after run of step 'StepHTTPServer'. Press enter to continue.
2020/01/21 10:12:40 ui: ==> hyperv-vmcx: Creating switch 'vSwitch' if required...
2020/01/21 10:12:41 ui: ==> hyperv-vmcx:     switch 'vSwitch' already exists. Will not delete on cleanup...
2020/01/21 10:12:41 ui: ask: ==> hyperv-vmcx: Pausing after run of step 'StepCreateSwitch'. Press enter to continue.
2020/01/21 10:13:30 ui: ==> hyperv-vmcx: Cloning virtual machine...
2020/01/21 10:13:30 packer.exe plugin: No existing virtual harddrive, not attaching.
2020/01/21 10:36:34 ui error: ==> hyperv-vmcx: Error cloning virtual machine: PowerShell error: Hyper-V\Set-VMMemory : Failed to modify device 'Memory'.
==> hyperv-vmcx: 'Win10Ent2019LTSC_GOLD' failed to modify device 'Memory'. (Virtual machine ID A93E429E-52FF-4879-8699-B58F9CEC4E55)
==> hyperv-vmcx: At C:\Users\usx00014\AppData\Local\Temp\2\powershell155259968.ps1:44 char:1
==> hyperv-vmcx: + Hyper-V\Set-VMMemory -VM $compatibilityReport.VM -StartupBytes $memor ...
==> hyperv-vmcx: + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
==> hyperv-vmcx:     + CategoryInfo          : InvalidOperation: (:) [Set-VMMemory], VirtualizationException
==> hyperv-vmcx:     + FullyQualifiedErrorId : InvalidState,Microsoft.HyperV.PowerShell.Commands.SetVMMemory
2020/01/21 10:36:34 ui: ==> hyperv-vmcx: keep_registered set. Skipping unregister/deletion of VM.
2020/01/21 10:36:34 ui: ask: ==> hyperv-vmcx: Pausing before cleanup of step 'StepCreateSwitch'. Press enter to continue.
2020/01/21 10:36:39 ui: ask: ==> hyperv-vmcx: Pausing before cleanup of step 'StepHTTPServer'. Press enter to continue.
2020/01/21 10:36:40 ui: ask: ==> hyperv-vmcx: Pausing before cleanup of step 'StepCreateFloppy'. Press enter to continue.
2020/01/21 10:36:40 ui: ask: ==> hyperv-vmcx: Pausing before cleanup of step 'StepOutputDir'. Press enter to continue.
2020/01/21 10:36:41 ui: ==> hyperv-vmcx: Deleting output directory...
2020/01/21 10:36:41 ui: ask: ==> hyperv-vmcx: Pausing before cleanup of step 'StepCreateBuildDir'. Press enter to continue.
2020/01/21 10:36:41 ui: ==> hyperv-vmcx: Deleting build directory...
2020/01/21 10:36:43 ui error: ==> hyperv-vmcx: Error deleting build directory: remove C:\Users\usx00014\AppData\Local\Temp\2\hyperv413148475\Virtual Machines\A93E429E-52FF-4879-8699-B58F9CEC4E55.vmcx: The process cannot access the file because it is being used by another process.
2020/01/21 10:36:43 [INFO] (telemetry) ending hyperv-vmcx
2020/01/21 10:36:43 ui error: Build 'hyperv-vmcx' errored: Error cloning virtual machine: PowerShell error: Hyper-V\Set-VMMemory : Failed to modify device 'Memory'.
'Win10Ent2019LTSC_GOLD' failed to modify device 'Memory'. (Virtual machine ID A93E429E-52FF-4879-8699-B58F9CEC4E55)
At C:\Users\usx00014\AppData\Local\Temp\2\powershell155259968.ps1:44 char:1
+ Hyper-V\Set-VMMemory -VM $compatibilityReport.VM -StartupBytes $memor ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Set-VMMemory], VirtualizationException
    + FullyQualifiedErrorId : InvalidState,Microsoft.HyperV.PowerShell.Commands.SetVMMemory
2020/01/21 10:36:43 Waiting on builds to complete...
2020/01/21 10:36:43 machine readable: error-count []string{"1"}
2020/01/21 10:36:43 ui error: 
==> Some builds didn't complete successfully and had errors:
2020/01/21 10:36:43 machine readable: hyperv-vmcx,error []string{"Error cloning virtual machine: PowerShell error: Hyper-V\\Set-VMMemory : Failed to modify device 'Memory'.\r\n'Win10Ent2019LTSC_GOLD' failed to modify device 'Memory'. (Virtual machine ID A93E429E-52FF-4879-8699-B58F9CEC4E55)\r\nAt C:\\Users\\usx00014\\AppData\\Local\\Temp\\2\\powershell155259968.ps1:44 char:1\r\n+ Hyper-V\\Set-VMMemory -VM $compatibilityReport.VM -StartupBytes $memor ...\r\n+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n    + CategoryInfo          : InvalidOperation: (:) [Set-VMMemory], VirtualizationException\r\n    + FullyQualifiedErrorId : InvalidState,Microsoft.HyperV.PowerShell.Commands.SetVMMemory"}
2020/01/21 10:36:43 ui error: --> hyperv-vmcx: Error cloning virtual machine: PowerShell error: Hyper-V\Set-VMMemory : Failed to modify device 'Memory'.
'Win10Ent2019LTSC_GOLD' failed to modify device 'Memory'. (Virtual machine ID A93E429E-52FF-4879-8699-B58F9CEC4E55)
At C:\Users\usx00014\AppData\Local\Temp\2\powershell155259968.ps1:44 char:1
+ Hyper-V\Set-VMMemory -VM $compatibilityReport.VM -StartupBytes $memor ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Set-VMMemory], VirtualizationException
    + FullyQualifiedErrorId : InvalidState,Microsoft.HyperV.PowerShell.Commands.SetVMMemory
2020/01/21 10:36:43 ui: 
==> Builds finished but no artifacts were created.
2020/01/21 10:36:43 [INFO] (telemetry) Finalizing.
2020/01/21 10:36:44 waiting for all plugin processes to complete...
2020/01/21 10:36:44 C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe: plugin process exited
2020/01/21 10:36:44 C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe: plugin process exited
2020/01/21 10:36:44 C:\Users\usx00014\Downloads\packer_1.5.1_windows_amd64\packer.exe: plugin process exited
```

**Comment by SwampDragons** [comment]
Is the vm in a "saved" state before you run the Packer builder? Have you tried actually powering it off fully and rerunning?

**Comment by xalaliah** [comment]
The exported VM is fully shut down and powered off when I run the script. 
I believe that when Packer clones from this VM, the cloned VM that Packer creates is in a "saved" state when it is trying to configure memory.

**Comment by xalaliah** [comment]
I'm not quite sure what happened here. This remained an issue for the last three weeks. 
For the heck of it, I decided to try building again this morning and it worked all the way through.
To make sure it wasn't a fluke, I built several more times with changes to `packer.json` but it has worked every single time.
I can't pinpoint what fixed it. The host has been pretty much untouched (has only been rebooted. No updates, installations, etc have been applied) since this issue was occurring, and I can't reproduce the issue anymore.
Closing since this seemed to be a weird anomaly with my host. I'll reopen if it occurs again.
